operators/pkg/instautoctrl/inactivity.go

```diff
@@ -0,0 +1,559 @@
+// Copyright 2020-2025 Politecnico di Torino
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package instautoctrl contains the controller for Instance Inactive Termination
+package instautoctrl
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/trace"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	clv1alpha2 "github.com/netgroup-polito/CrownLabs/operators/api/v1alpha2"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/forge"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/utils"
+)
+
+// InstanceInactiveTerminationReconciler watches for instances to be terminated.
+type InstanceInactiveTerminationReconciler struct {
+	client.Client
+	EventsRecorder            record.EventRecorder
+	Scheme                    *runtime.Scheme
+	NamespaceWhitelist        metav1.LabelSelector
+	StatusCheckRequestTimeout time.Duration
+	InstanceMaxNumberOfAlerts int
+	MailClient                *utils.MailClient
+	PrometheusURL             string
+	InactivityInterval        time.Duration
+	// This function, if configured, is deferred at the beginning of the Reconcile.
+	// Specifically, it is meant to be set to GinkgoRecover during the tests,
+	// in order to lead to a controlled failure in case the Reconcile panics.
+	ReconcileDeferHook func()
+}
+
+var alertAnnotation = "crownlabs.polito.it/number-alerts-sent"
+var lastActivityAnnotation = "crownlabs.polito.it/last-activity"
```

Comment:
there should be a labels.go in forge pkg, see how labels are usually applied and try to use the same way